java/patterns/src/main/java/dev/restate/patterns/Compensations.java

```diff
@@ -0,0 +1,94 @@
+package dev.restate.patterns;
+
+import static dev.restate.patterns.compensations.generated.Proto.*;
+
+import dev.restate.patterns.compensations.generated.*;
+import dev.restate.sdk.RestateContext;
+import dev.restate.sdk.common.TerminalException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * -------------------------------------- Compensations --------------------------------------
+ * Restate guarantees that your invocations are run to completion no matter what happens. However,
+ * in case you want your services to be able to respond with exceptions as part of the normal
+ * control flow (i.e. denoting a negative outcome) you can let the callee throw a
+ * `TerminalException`. If the caller has completed previous steps before receiving the
+ * `TerminalException`, it will have to undo those steps in order to leave the system in a
+ * consistent state. The way to undo steps is by registering compensations and running them for all
+ * completed steps. With Restate, this can be naturally expressed as part of the service code.
+ * Moreover, Restate runs the compensations with the same guarantees as the service code, i.e. it
+ * executes them durably until they complete, which guarantees that the system will be left in a
+ * consistent state.
+ *
+ * <p>The very same mechanism of undoing completed steps via compensations is also needed if you
+ * want to gracefully cancel an invocation, i.e. because it is no longer needed or stuck, and leave
+ * the system in a consistent state. The way a graceful cancellation manifests is by throwing a
+ * `TerminalError` at the next Restate API call. In fact, Restate starts cancelling your invocation
+ * call tree starting with the leave invocations first and then propagating the cancellation error
```

Comment:
Minor typo! At least I _think_ this is what you mean :-)

```suggestion
 * call tree starting with the leaf invocations first and then propagating the cancellation error
```

Reply:
Thanks for catching this. It is indeed leaf.